### PR TITLE
replace the default ThreadPoolExecutor with gevent.threadpool.ThreadPoolExecutor

### PR DESCRIPTION
**Pull request by allen-munsch**
I have no idea if this will work, but maybe it will?

Here's a bit of my research on what I was thinking: https://github.com/rednaks/django-async-orm/discussions/9

```
web_1         | DEBUG:root:concurrent.futures.Future is expected, got <gevent.threadpool._FutureProxy object at 0x7fcabce2e9d0>
web_1         | DEBUG:root:concurrent.futures.Future is expected, got <gevent.threadpool._FutureProxy object at 0x7fcabf109bb0>
web_1         | DEBUG:root:concurrent.futures.Future is expected, got <gevent.threadpool._FutureProxy object at 0x7fcabf109250>
web_1         | DEBUG:root:concurrent.futures.Future is expected, got <gevent.threadpool._FutureProxy object at 0x7fcabc5654f0>
web_1         | DEBUG:root:concurrent.futures.Future is expected, got <gevent.threadpool._FutureProxy object at 0x7fcabd7846d0>
^CGracefully stopping... (press Ctrl+C again to force)
Stopping benchmark-django-fastapi_web_1      ... done
Stopping benchmark-django-fastapi_postgres_1 ... done

```

looks kind of close ... although something is expecting `concurrent.futures.Future` instead of also allowing `gevent.threadpool._FutureProxy`

update, a recursion error of some sort

[recursion.error.txt](https://github.com/rednaks/django-async-orm/files/6558492/recursion.error.txt)

### CODE REVIEW

File: django_async_orm/utils.py [hunk 1]
```diff
@@ -12,9 +12,9 @@ def mixin_async_manager_factory(model):
     """
 
     base_manager_cls = model.objects.__class__
-    mixin_async_manager = type(f'MixinAsync{base_manager_cls.__name__}', (AsyncManager, base_manager_cls), dict())
-
-    return mixin_async_manager
+    if 'MixinAsyncUserManager' != f'MixinAsync{base_manager_cls.__name__}':
```
Comment (rednaks) [inline django_async_orm/utils.py:15]: is this happening because you have custom user model that inherited from `User` or `Abstract*User` ?

**Comment by rednaks** [comment]
Thank you for the PR @allen-munsch ! 

Reading the doc about  `gevent.threadpool.ThreadPoolExecutor` seems to me a good idea to use the native threads if they are patched by gevent. 

>  class ThreadPoolExecutor(*args, **kwargs)[source]
>
>    Bases: concurrent.futures.thread.ThreadPoolExecutor
>
>    A version of concurrent.futures.ThreadPoolExecutor that always uses native threads, even when threading is monkey-patched.
>
>   The Future objects returned from this object can be used with gevent waiting primitives like gevent.wait().

But at the same time, I'm wondering why would I use `gevent` with asyncio ? I believe dev should use either `gevent` or `asyncio` It's already risky to use gevent to patch everything since we don't have a hand over the cooperation between the tasks.

On the other hand, if everything is already patched with `gevent` would it still make sense to use `asyncio` ?

It might be a good idea at the end but it's a complete obscure to me at the moment.

File: django_async_orm/utils.py (hunk 1 above)
Comment (rednaks) [reply django_async_orm/utils.py:15]: Would it make more sense to test if the parent model have been already patched:
```python
if base_manager_cls.__name__.startswith('MixinAsync'):
  ...
```

Comment (rednaks) [reply django_async_orm/utils.py:15]: This should be in a different PR so we can merge it independently

Comment (allen-munsch) [reply django_async_orm/utils.py:15]: @rednaks 

maybe from doing `User = get_user_model()`

also : https://github.com/rednaks/django-async-orm/pull/10

**Comment by allen-munsch** [comment]
> Thank you for the PR @allen-munsch !
> 
> Reading the doc about `gevent.threadpool.ThreadPoolExecutor` seems to me a good idea to use the native threads if they are patched by gevent.
> 
> > class ThreadPoolExecutor(*args, **kwargs)[source]
> > Bases: concurrent.futures.thread.ThreadPoolExecutor
> > A version of concurrent.futures.ThreadPoolExecutor that always uses native threads, even when threading is monkey-patched.
> > The Future objects returned from this object can be used with gevent waiting primitives like gevent.wait().
> 
> But at the same time, I'm wondering why would I use `gevent` with asyncio ? I believe dev should use either `gevent` or `asyncio` It's already risky to use gevent to patch everything since we don't have a hand over the cooperation between the tasks.
> 
> On the other hand, if everything is already patched with `gevent` would it still make sense to use `asyncio` ?
> 
> It might be a good idea at the end but it's a complete obscure to me at the moment.

I was looking into it, and it seemed like the gevent.threadpool.ThreadPoolExcutor was basically a wrapper around the concurrent.futures.ThreadPoolExecutor, where a few of the attributes, `_condition`, `_state`, and a few of the methods do the gevent magic.

The main reason I was looking into this is this. 

Say for example a shop is using django 3.2 but everything is all WSGI with gevent/eventlet, and psycogreen.

What steps would need to be taken in order to begin adding in some extra ASGI stuff?

- FastApi is really cool, but it'd be a real pain to re-write the entire django orm model schema for some big old legacy code base
- The approach FastAPI suggests in the docs is to wrap WSGI and mount it to ASGI
- I've tried that, and since everything was asyncio, all of the psycogreen/eventlet/gevent database stuff became an issue.
- So I was thinking, another approach might be to  wrap an ASGI app and mount it to WSGI
- re-using the django orm in the newer ASGI app with sync_to_async

I guess the main question I'm trying to answer is: 

```
Will using concurrent.futures.ThreadPoolExecutor as the default for sync_to_async in a monkey patched gevent WSGI app block the database connections? 

Eventually causing a max connection issue?
```

This PR is more of an experimental workspace for : https://github.com/rednaks/django-async-orm/discussions/9

**Comment by allen-munsch** [comment]
closing in favor of :  #9

**Comment by allen-munsch** [comment]
https://github.com/django/asgiref/issues/264

**Pull request closed without merging**